gadget/mkfs.go

```diff
@@ -0,0 +1,121 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// MkfsExt4 creates an EXT4 filesystem in given image file, with an optional
+// filesystem label, and populates it with the contents of provided root
+// directory.
+func MkfsExt4(img, label, contentsRootDir string) error {
+	// taken from ubuntu-image
+	// TODO: support e2fsprogs 1.42 without -d in Ubuntu 16.04
+	mkfsArgs := []string{
+		"mkfs.ext4",
+		// default usage type
+		"-T", "default",
+		// disable metadata checksum
+		// XXX: why?
+		"-O", "-metadata_csum",
```

Comment:
I think we should clarify with foundations (cc @vorlonofportland) why ubuntu-image is using these defaults. It seems beneficial to use metadata checksums so having at least a comment about the "why disabled" here would be good.

Reply:
@mvo5 metadata checksums were unsupported by the e2fsprogs in Ubuntu 16.04 (and therefore in Ubuntu Core 16).  Creating filesystems with metadata checksums (which is the default with the e2fsprogs in Ubuntu 18.04!) would directly lead to a boot failure.

Reply:
@vorlonofportland thanks for the input. I'll update the comment accordingly.